File: 2022/go/4.go
package main

import (
	"fmt"
	"os"
	"strings"
	"strconv"
)

func main() {

	input, err := os.ReadFile("../4.txt")
	if err != nil {
		fmt.Printf("Could not read the file due to this %s error \n", err)
	}
	sc := string(input)
	lines := strings.Split(sc, "\n")

	var score_part1 int = 0
	var score_part2 int = 0

	for i := 0; i < len(lines); i++ {
		pair := strings.Split(lines[i], ",")
		first := strings.Split(pair[0], "-")
		second := strings.Split(pair[1], "-")

		first_start,_ :=  strconv.Atoi(first[0])
        first_end,_ := strconv.Atoi(first[1])
        second_start,_ := strconv.Atoi(second[0]) 
        second_end,_ := strconv.Atoi(second[1]) 

		if (first_start <= second_start && first_end >= second_end) || ( second_start <= first_start && second_end >= first_end ) {
            score_part1 += 1;
        }

        if (first_start <= second_end && second_start <= first_end) {
            score_part2 += 1;
        }
	}

	
	fmt.Printf("Score 1: %v\n", score_part1)	
	fmt.Printf("Score 2: %v\n", score_part2)	
}

File: 2022/rust/4.rs
use std::fs;
use std::collections::HashSet;

fn main() {
    let input = fs::read_to_string("../4.txt").unwrap();
    let lines : Vec<&str> = input.split("\n").collect();

    let mut score_part1 : u32 = 0;
    let mut score_part2 : u32 = 0;

    // Part 1
    for line in &lines {
        let pair : Vec<&str> = line.trim().split(",").collect();
        let first : Vec<&str> = pair[0].split("-").collect();
        let second : Vec<&str> = pair[1].split("-").collect();

        let first_start = first[0].parse::<i32>().unwrap();
        let first_end = first[1].parse::<i32>().unwrap();
        let second_start = second[0].parse::<i32>().unwrap();
        let second_end = second[1].parse::<i32>().unwrap();
        
        if (first_start <= second_start && first_end >= second_end) || ( second_start <= first_start && second_end >= first_end ) {
            score_part1 += 1;
        }

        if (first_start <= second_end && second_start <= first_end) {
            score_part2 += 1;
        }
    }

    
    println!("Score 1: {}", score_part1);
    println!("Score 2: {}", score_part2);
    
}